### PR TITLE
[WIP] Move to pull-streams

### PR DESCRIPTION
**Pull request by dignifiedquire**

### CODE REVIEW

File: examples/dialer.js [hunk 1]
```diff
@@ -1,26 +1,25 @@
...
+const pull = require('pull-stream')
+const toPull = require('stream-to-pull-stream')
 const libp2pSPDY = require('../src')
 
 const socket = tcp.connect(9999)
-const muxer = libp2pSPDY(socket, false)
+const muxer = libp2pSPDY.dial(toPull(socket))
 
 muxer.on('stream', (stream) => {
   console.log('-> got new muxed stream')
```
Comment (diasdavid) [inline examples/dialer.js]: Looks like I had forgot to remove this one.

File: package.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "name": "libp2p-spdy",
   "version": "0.8.1",
   "description": "SPDY 3.1 implementation wrapper that is compatible with libp2p Stream Muxer expected interface",
-  "main": "lib/index.js",
+  "main": "src/index.js",
```
Comment (diasdavid) [inline package.json]: - [ ] change before merge

File: src/index.js [hunk 3]
```diff
@@ -1,70 +1,20 @@
...
-  proxyMuxer.multicodec = exports.multicodec
-  return proxyMuxer
+  return new Muxer(conn, spdyMuxer)
 }
 
-exports.multicodec = '/spdy/3.1.0'
-
-function noop () {}
+exports.multicodec = SPDY_CODEC
+exports.dial = (conn) => create(conn, false)
```
Comment (diasdavid) [inline src/index.js]: `s/dial/dialer`
`s/listen/listener`

File: src/index.js [hunk 4]
```diff
@@ -1,70 +1,20 @@
...
+exports.listen = (conn) => create(conn, true)
```
Comment (diasdavid) [inline src/index.js]: This also changes the interface btw, must update https://github.com/libp2p/interface-stream-muxer

**Comment by diasdavid** [comment]
Have you run the https://github.com/libp2p/interface-stream-muxer tests? I would love to see a `time` benchmark between using regular streams and pull streams

Let's also have tests for:

- [ ] destroy underlying pull stream propagates the close and error to all the streams above
- [ ] close one of the muxed streams doesn't close others that are open
- [ ] close on spdy doesn't close until all the streams that are being muxed are closed too

**Comment by dignifiedquire** [comment]
> Have you run the https://github.com/libp2p/interface-stream-muxer tests? I would love to see a time benchmark between using regular streams and pull streams

I haven't run them yet, I will go and update them + move them to `mocha` as that way they will be always run with the main test suite.

**Comment by diasdavid** [comment]
> they will be always run with the main test suite.

They can be run on release, but not every single as they take some quite of time to finish.

**Comment by diasdavid** [comment]
Getting consistent:

```
messages 1446
endStreams 14
streamsCreated 100
```

on the:

```
    it.only('100 stream with 100 msg', (done) => {
      spawn(muxer, 100, 100, done)
    })
```

**Comment by diasdavid** [comment]
More detailed:

```
 # 1 1
messages 1
endStreams 1
streamsCreated 1

 # 1 10
messages 10
endStreams 1
streamsCreated 1

 # 1 100
messages 100
endStreams 1
streamsCreated 1

 # 10 1
messages 10
endStreams 10
streamsCreated 10

 # 10 10
messages 100
endStreams 10
streamsCreated 10

 # 10 100
messages 1000
endStreams 10
streamsCreated 10

 # 100 1
messages 100
endStreams 100
streamsCreated 100

 # 100 10
messages 1000
endStreams 100
streamsCreated 100

 # 100 100
messages 1446
endStreams 14
streamsCreated 100
```

**Comment by diasdavid** [comment]
@dignifiedquire could you push an update to this branch with the remove of the throttle on stress tests and the pointer to the forked version of spdy?

**Comment by dignifiedquire** [comment]
@diasdavid done

**Comment by diasdavid** [comment]
@dignifiedquire still missing these tests:

https://github.com/libp2p/js-libp2p-spdy/pull/16#issuecomment-240680117

**Comment by dignifiedquire** [comment]
> Have you run the https://github.com/libp2p/interface-stream-muxer tests? I would love to see a time benchmark between using regular streams and pull streams

It's not using the same test system anymore, but here are the before and after times

```
# current master
total:     130127
passing:   130127
duration:  11.4s
npm run compliance  14.52s user 1.42s system 131% cpu 12.108 total
```

```
# pull
npm run test:node  5.70s user 0.55s system 107% cpu 5.824 total
```

**Comment by dignifiedquire** [comment]
@diasdavid could you write some pseudo code for the tests you are suggesting, I'm not entirely sure what you want.

**Comment by diasdavid** [comment]
@dignifiedquire added more info there - https://github.com/libp2p/js-libp2p-spdy/pull/16#issuecomment-240680117

File: test/spdy-over-tcp.node.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  let dialer
+
+  let tcp
+  let mh = multiaddr('/ip4/127.0.0.1/tcp/9090')
+
+  before(() => {
+    tcp = new Tcp()
+  })
+
+  it('attach to a websocket, as listener', (done) => {
```
Comment (diasdavid) [inline test/spdy-over-tcp.node.js]: should be to a tcp socket

File: test/spdy-over-tcp.node.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    const tcpListener = tcp.createListener((socket) => {
+      expect(socket).to.exist
+      listener = spdy.listen(socket)
+      expect(listener).to.exist
+    })
+
+    tcpListener.listen(mh, done)
+  })
+
+  it('attach to a websocket, as dialer', (done) => {
```
Comment (diasdavid) [inline test/spdy-over-tcp.node.js]: same as above

**Comment by diasdavid** [comment]
Last things for the merge

- [ ] https://github.com/libp2p/js-libp2p-spdy/pull/16#issuecomment-240680117
- [ ] https://github.com/libp2p/js-libp2p-spdy/pull/16#discussion_r75280473
- [ ] https://github.com/libp2p/js-libp2p-spdy/pull/16#discussion_r75280560
- [ ] Update readme with documentation about pull-streams

**Comment by dignifiedquire** [comment]
> close on spdy doesn't close until all the streams that are being muxed are closed too

test exists, but fails, not sure why though :(

**Comment by diasdavid** [comment]
@dignifiedquire you mentioned that spdy is build to destroy all the muxedConns if we call it to end. If that is the case, then it should close all the streams and not wait for them to finish.

**Comment by dignifiedquire** [comment]
> you mentioned that spdy is build to destroy all the muxedConns if we call it to end

This was a misunderstanding from my side. Calling `.end` will

- Send `goaway` frame
-  Error all newly created streams with `New stream after goaway`
- Attach an optional `close` handler, which is triggered once all open streams are closed

**Comment by diasdavid** [comment]
awesome, so it doesn't abruptly all flowing streams, that is what I thought :)

**Comment by dignifiedquire** [comment]
> let's add to the list the same test, but make it happen where data is still being sent.

that's what the test currently does, send data on 5 streams delayed, and trigger `.end` before they are finished

**Pull request merged**